source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,333 @@
+  // for kDNSServiceFlagsShareConnection in dns_sd.h, and copied (and edited) in this implementation
+  // where relevant.
+  auto error = DNSServiceCreateConnection(&main_sd_ref_);
+  RELEASE_ASSERT(!error, "error in DNSServiceCreateConnection");
+
+  auto fd = DNSServiceRefSockFD(main_sd_ref_);
+  RELEASE_ASSERT(fd != -1, "error in DNSServiceRefSockFD");
+  ENVOY_LOG(debug, "DNS resolver has fd={}", fd);
+
+  sd_ref_event_ = dispatcher_.createFileEvent(
+      fd,
+      // note: Event::FileTriggerType::Level is used here to closely resemble the c-ares
+      // implementation in dns_impl.cc.
+      [this](uint32_t events) { onEventCallback(events); }, Event::FileTriggerType::Level,
+      Event::FileReadyType::Read);
+  sd_ref_event_->setEnabled(Event::FileReadyType::Read);
+}
+
+void AppleDnsResolverImpl::onEventCallback(uint32_t events) {
+  ENVOY_LOG(debug, "DNS resolver file event");
+  ASSERT(events & Event::FileReadyType::Read);
+  DNSServiceProcessResult(main_sd_ref_);
+}
+
+ActiveDnsQuery* AppleDnsResolverImpl::resolve(const std::string& dns_name,
+                                              DnsLookupFamily dns_lookup_family,
+                                              ResolveCb callback) {
+  ENVOY_LOG(debug, "DNS resolver resolve={}", dns_name);
+  std::unique_ptr<PendingResolution> pending_resolution(
+      new PendingResolution(*this, callback, dispatcher_, main_sd_ref_, dns_name));
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    ENVOY_LOG(error, "DNS resolver error in dnsServiceGetAddrInfo for {}", dns_name);
+    return nullptr;
+  }
+
+  // If the query was synchronously resolved, there is no need to return the query.
+  if (pending_resolution->synchronously_completed_) {
+    return nullptr;
+  }
+
+  pending_resolution->owned_ = true;
+  return pending_resolution.release();
+}
+
+void AppleDnsResolverImpl::addPendingQuery(PendingResolution* query) {
+  ASSERT(queries_with_pending_cb_.count(query) == 0);
+  queries_with_pending_cb_.insert(query);
+}
+
+void AppleDnsResolverImpl::removePendingQuery(PendingResolution* query) {
+  auto erased = queries_with_pending_cb_.erase(query);
+  ASSERT(erased == 1);
+}
+
+void AppleDnsResolverImpl::flushPendingQueries(const bool with_error) {
+  ENVOY_LOG(debug, "DNS Resolver flushing {} queries", queries_with_pending_cb_.size());
+  for (std::set<PendingResolution*>::iterator it = queries_with_pending_cb_.begin();
+       it != queries_with_pending_cb_.end(); ++it) {
+    auto query = *it;
+    try {
+      ASSERT(query->pending_cb_);
+      query->callback_(query->pending_cb_->status_, std::move(query->pending_cb_->responses_));
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(critical, "EnvoyException in DNSService callback: {}", e.what());
+      query->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
```

Comment:
Since this always happens on the same thread, are these posts necessary still. Is it to avoid throwing through the API callback? Can you add comments?

Reply:
whoops sorry I actually removed but the commit got snagged in pre-push hook. Will go up in a sec